### PR TITLE
[TRTLLM-12288][feat] Support Nemotron-H nvfp4 ckpt on Hopper

### PR DESCRIPTION
**Pull request by JadoTu**
## Summary by CodeRabbit

* **New Features**
  * Added W4A16 NVFP4 quantization support with automatic fallback handling for Hopper architecture.

* **Tests**
  * Added accuracy validation test for W4A16 NVFP4 quantization on 4-GPU Hopper systems.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14775?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

## Description
Enables NVFP4 Nemotron-H checkpoints to run on Hopper GPUs, which lack a native NVFP4 tensor-core GEMM. We add a W4A16 path that loads the NVFP4 weights and dequantizes them on-the-fly per forward step via Triton kernel. CUDA-graph capturable; Blackwell paths are untouched.

1. This method will fallback different moe backends to one, which is a new path for nvfp4 on hopper.
2. The dequantization of nvfp4 is done by triton, which needs further improvement.
3. Functional tests have passed with Nemotron ultra/super on h100/h200, enabling MTP.
4. The perf now is not tuned. Here is a example of ultra model running the full dataset of gsm8k on 8xh100: 13 mins.
5. A new ci test is added.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 1]
```diff
@@ -431,10 +438,14 @@ def __init__(
 
         quant_mode = (model_config.quant_config.quant_mode
                       if model_config.quant_config is not None else None)
-        self.is_nvfp4 = quant_mode is not None and quant_mode.has_nvfp4()
+        # We don't use the RMSNorm+NVFP4 on SM < 100
+        _has_fp4_hw = get_sm_version() >= 100
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_nemotron_h.py:442]: only feasible on >=100 while <120,

only b200/b300 sm version has such support.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 2]
```diff
@@ -754,6 +770,93 @@ def forward(
         return hidden_states
 
 
+def _force_moe_backend_for_w4a16_on_hopper(
+        model_config: NemotronHModelConfig) -> None:
+    """SM<100 + NVFP4: force ``moe_backend=CUTLASS`` (only backend with the
+    W4A16 fallback) and disable attention FP4 output fusion.
+    """
+    if get_sm_version() >= 100:
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_nemotron_h.py:778]: same sm version checking here.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 3]
```diff
@@ -7141,6 +7141,35 @@ def test_nvfp4_4gpus_online_eplb(self, moe_backend):
...
+        pytorch_config = dict(cuda_graph_config=cuda_graph_config)
+        with LLM(
+                model_path,
+                kv_cache_config=kv_cache_config,
+                max_batch_size=max_batch_size,
+                tensor_parallel_size=4,
+                speculative_config=mtp_config,
+                **pytorch_config,
+        ) as llm:
+            task = GSM8K(self.MODEL_NAME)
```
Comment (Wanli-Jiang) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py:7169]: can we also add mmlu checking?

File: tensorrt_llm/_torch/modules/linear.py [hunk 4]
```diff
@@ -1861,6 +1861,87 @@ def post_load_weights(self, module: Linear):
...
+            original_shape = input.shape
+            input = input.reshape(-1, input.shape[-1])
+
+        # NVFP4_AWQ pre_quant_scale (mirrors parent's _input_prepare branch).
+        if module.pre_quant_scale is not None:
+            assert input.dtype == module.pre_quant_scale.dtype, (
+                "Input dtype and pre_quant_scale dtype must match")
+            input = input * module.pre_quant_scale
+
+        from tensorrt_llm._torch.modules.fused_moe.triton_dequant_nvfp4 import \
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/linear.py:1909]: could we move this import to the begining of the file?

File: tensorrt_llm/_torch/modules/linear.py [hunk 5]
```diff
@@ -1861,6 +1861,87 @@ def post_load_weights(self, module: Linear):
...
+        if isinstance(input, (Fp4QuantizedTensor, tuple)):
+            raise RuntimeError(
+                "W4A16NVFP4LinearMethod: hp input required; disable upstream "
+                "FP4 fusion (e.g. TRTLLM_ENABLE_ATTENTION_NVFP4_OUTPUT=0)")
+
+        ## FP8 input from upstream FMHA pre-quant: invert by / module.inv_input_scale.
+        if input.dtype == torch.float8_e4m3fn:
+            assert module.inv_input_scale is not None, \
+                "W4A16NVFP4LinearMethod: FP8 input requires static inv_input_scale"
+            input = (input.to(module.dtype) / module.inv_input_scale).to(
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/linear.py:1895]: should we add a check to input or inv_input_scale, to avoid zero_division?

File: tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
+# SPDX-FileCopyrightText: Copyright (c) 2022-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py:1]: `2022-2026` -> `2026`

File: tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+W4A16NVFP4CutlassFusedMoEMethod). Static shapes -> CUDA-graph capturable.
+
+Pipeline: scatter active_mask[E] from routing -> static (E, N/BN, K/BK) grid
+-> inactive blocks early-return, active blocks dequant their tile. Downstream
+fused_moe only reads rows in token_selected_experts, so leaving inactive
+rows uninitialized is safe.
+"""
+
+import torch
+import triton  # type: ignore[import]
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py:13]: curious why we need such `# type: ignore[import]`?

File: tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    )
+
+    N, K_packed = packed_weight.shape
+    K = K_packed * 2
+    device = packed_weight.device
+
+    # Reshape (possibly flat) scale to its 2D [pad_rows, pad_cols] form so
+    # the kernel can use ``scale.stride(0)`` directly.
+    if weight_scale.dim() == 1:
+        from tensorrt_llm.quantization.utils.fp4_utils import pad_up
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py:303]: move this import to the file head.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py (hunk 1 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/models/modeling_nemotron_h.py:442]: Nice reminder, will fix.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 3 above)
Comment (JadoTu) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py:7169]: The perf for this hopper flow is not that good now. I think more tests will slow down all ci processes.

File: tensorrt_llm/_torch/modules/linear.py (hunk 4 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/linear.py:1909]: Will do.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 9]
```diff
@@ -2892,6 +2892,84 @@ def process_weights_after_loading(self, module: torch.nn.Module):
...
+    into a static [E_total, N, K] workspace, then runs the bf16 ``fused_moe``.
+    """
+
+    def process_weights_after_loading(self, module: torch.nn.Module):
+        super().process_weights_after_loading(module)
+
+        # Scale buffer: int32-packed FP8, viewed as uint8 has shape
+        # [E, pad_up(N, 128), pad_up(K/sf_vec, 4)] -- the 3D layout
+        # block_scale_interleave_reverse accepts.
+        def _unswizzle_inplace(scale_param: torch.nn.Parameter):
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py:2911]: one comment from AI agent:
3D tensor passed to block_scale_interleave_reverse in MoE unswizzle
In W4A16NVFP4CutlassFusedMoEMethod.process_weights_after_loading, the _unswizzle_inplace helper passes a 3D [E, pad_rows, pad_cols] tensor to torch.ops.trtllm.block_scale_interleave_reverse (quantization.py:2915). All other usages of this op in the codebase (e.g., linear.py:1879, and the parent class _interleave_w3_w1_weight_scale at quantization.py:2731) operate on 2D tensors. If this C++ op does not support batched 3D input, the unswizzled scales would be incorrect, silently corrupting MoE weight dequantization. The variables E, pad_rows, pad_cols are also computed but never used (dead code at line 2913-2914). Worth verifying that the C++ implementation handles the batch dimension correctly.

File: tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py (hunk 8 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py:303]: will do

**Comment by JadoTu** [comment]
Thanks for you review @Wanli-Jiang . I will collect all the changes and apply them in next follow-up PR. The new model will be released soon so that I don't want to push new commits and rerun the CI. In this way we can first have this functional flow on the new model. I will paste the new PR once I get all changes.

File: tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py (hunk 7 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py:13]: I would like to say it's ai's mistake. I will remove it.

**Review by Wanli-Jiang — APPROVED** [review]

**Review by HuiGao-NV — APPROVED** [review]
LGTM

File: tests/integration/test_lists/test-db/l0_dgx_h100.yml [hunk 10]
```diff
@@ -102,6 +102,7 @@ l0_dgx_h100:
   - accuracy/test_llm_api_pytorch.py::TestNemotronV3Super::test_fp8_4gpus[attention_dp_off-cpp_mamba_cache]
   - accuracy/test_llm_api_pytorch.py::TestNemotronV3Super::test_fp8_4gpus[attention_dp_on-python_mamba_cache]
   - accuracy/test_llm_api_pytorch.py::TestNemotronV3Super::test_fp8_4gpus[attention_dp_on-cpp_mamba_cache]
+  - accuracy/test_llm_api_pytorch.py::TestNemotronV3Super::test_nvfp4_4gpus_hopper_w4a16
```
Comment (xinhe-nv) [inline tests/integration/test_lists/test-db/l0_dgx_h100.yml:105]: please add this case into qa test list.

**Review by xinhe-nv — APPROVED** [review]

File: tests/integration/test_lists/test-db/l0_dgx_h100.yml (hunk 10 above)
Comment (JadoTu) [reply tests/integration/test_lists/test-db/l0_dgx_h100.yml:105]: Will do this in the next PR.

File: tensorrt_llm/_torch/modules/linear.py (hunk 5 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/linear.py:1895]: I think no need here. inv_input_scale = 448 / amax_input. For this to be zero, amax_input will need to be inf. It won't happen in a normal ckpt other than corruption.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py (hunk 9 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py:2911]: 3D support verified in cpp/tensorrt_llm/thop/fp4Op.cpp:254.
I will clean the dead code then.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 11]
```diff
@@ -754,6 +770,93 @@ def forward(
...
+    if not has_nvfp4:
+        return
+
+    # o_proj.has_nvfp4 stays True under W4A16 -- the property reads quant_config.
+    # Use the documented env override to keep attention output in bf16.
+    if os.environ.get("TRTLLM_ENABLE_ATTENTION_NVFP4_OUTPUT") != "0":
+        logger.warning(
+            f"Nemotron-H SM{get_sm_version()}: TRTLLM_ENABLE_ATTENTION_NVFP4_OUTPUT=0"
+        )
+        os.environ["TRTLLM_ENABLE_ATTENTION_NVFP4_OUTPUT"] = "0"
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tensorrt_llm/_torch/models/modeling_nemotron_h.py:797]: fix.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 12]
```diff
@@ -2892,6 +2892,84 @@ def process_weights_after_loading(self, module: torch.nn.Module):
...
+            sf_view = scale_param.data.view(float4_sf_dtype)
+            E, pad_rows, pad_cols = (sf_view.shape[0], sf_view.shape[1],
+                                     sf_view.shape[2])
+            linear = torch.ops.trtllm.block_scale_interleave_reverse(sf_view)
+            scale_param.data.view(float4_sf_dtype).copy_(linear)
+
+        _unswizzle_inplace(module.w3_w1_weight_scale)
+        _unswizzle_inplace(module.w2_weight_scale)
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py:2919]: fix.

File: tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+        out.stride(1),
+        # shapes
+        N,
+        K,
+        # constexpr
+        SF_VEC=sf_vec_size,
+        BLOCK_N=block_n,
+        BLOCK_K=block_k,
+    )
+    return out
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/fused_moe/triton_dequant_nvfp4.py:194]: fix

File: tensorrt_llm/_torch/modules/linear.py [hunk 14]
```diff
@@ -1861,6 +1861,87 @@ def post_load_weights(self, module: Linear):
...
+                module.dtype)
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tensorrt_llm/_torch/modules/linear.py:1896]: fix

**Comment by JadoTu** [comment]
The follow-up PR is: https://github.com/NVIDIA/TensorRT-LLM/pull/14862
